superduperdb/vector_search/atlas.py

```diff
@@ -0,0 +1,141 @@
+                    'numCandidates': n,
+                }
+            },
+            {'$addFields': {'score': {'$meta': 'vectorSearchScore'}}},
+        ]
+
+        cursor = self.index.aggregate(
+            Aggregate._prepare_pipeline(
+                pl,
+                self.db,
+                self.vi,
+            )
+        )
+
+        # Execute the query to get vector
+        return list(cursor)
+
+    def find_nearest_from_id(self, id, n=100, within_ids=None):
+        # TODO: Write logic to retrive vector from id.
+        h = self.index.find_one({'id': id})
+        return self.find_nearest_from_array(h, n=n, within_ids=within_ids)
+
+    def find_nearest_from_array(self, h, n=100, within_ids=None):
+        return self._find(h, n=n)
+
+    def add(self, items):
+        items = list(map(lambda x: x.to_dict(), items))
+        if not CFG.cluster.vector_search == CFG.data_backend:
+            self.index.insert_many(items)
+
+    def delete(self, items):
+        ids = list(map(lambda x: x.id, items))
+        if not CFG.cluster.vector_search == CFG.data_backend:
+            self.index.delete_many({'id': {'$in': ids}})
+
+    def _create_index(self, vector_index, db):
+        """
+        Create a vector index in the data backend if an Atlas deployment.
+
+        :param vector_index: vector index to create
+        """
+        collection = self.index.name
+        key = vector_index.indexing_listener.key
+        if re.match('^_outputs\.[A-Za-z0-9_]+\.[A-Za-z0-9_]+', key):
+            key = key.split('.')[1]
+        model = vector_index.indexing_listener.model.identifier
+        version = vector_index.indexing_listener.model.version
+        fields4 = {
+            str(version): [
+                {
+                    "dimensions": vector_index.dimensions,
+                    "similarity": vector_index.measure,
+                    "type": "knnVector",
+                }
+            ]
+        }
+        fields3 = {
+            model: {
+                "fields": fields4,
+                "type": "document",
+            }
+        }
+        fields2 = {
+            key: {
+                "fields": fields3,
+                "type": "document",
+            }
+        }
+        fields1 = {
+            "_outputs": {
+                "fields": fields2,
+                "type": "document",
+            }
+        }
+        index_definition = {
+            "createSearchIndexes": collection,
+            "indexes": [
+                {
+                    "name": vector_index.identifier,
+                    "definition": {
+                        "mappings": {
+                            "dynamic": True,
+                            "fields": fields1,
+                        }
+                    },
+                }
+            ],
+        }
+        print(json.dumps(index_definition, indent=2))
+        db.command(index_definition)
+
+    def create_index(self, vi: 'VectorIndex', db: 'Datalayer'):
```

Comment:
This function should be private and not called from outside.

In `__init__` we need a check: "does this index exist". If not, then create.

See the description in the issue: https://github.com/SuperDuperDB/superduperdb/issues/1605